openmc/element.py

```diff
@@ -102,9 +123,160 @@ def expand(self):
+                        i = min(list(range(len(nuclides_a))), key=lambda j: \
+                                abs(nuclides_a[j] - a))
+
+                        # Increment abundance of the nearest nuclide
+                        abundances[i] += abundance
+
+        # If a cross_section library is not present, expand the element into
+        # its natural nuclides
+        else:
+            for nuclide, abundance in sorted(NATURAL_ABUNDANCE.items()):
+                if re.match(r'{}\d+'.format(self.name), nuclide):
+                    nuclides.append(openmc.Nuclide(nuclide))
+                    abundances.append(abundance)
+
+        # Create a list of atomic masses
+        n_nuclides = len(nuclides)
+        atomic_masses = []
+        for nuclide in nuclides:
+            atomic_masses.append(openmc.data.atomic_mass(nuclide.name))
+
+        # Modify mole fractions if enrichment provided
+        if enrichment is not None:
+
+            # Get the indices for the uranium nuclides
+            for i,nuc in enumerate(nuclides):
+                if nuc.name == 'U234':
+                    u234 = i
+                elif nuc.name == 'U235':
+                    u235 = i
+                elif nuc.name == 'U238':
+                    u238 = i
+
+            # Calculate the mass fractions of isotopes
+            abundances[u234] = 0.008 * enrichment
+            abundances[u235] = enrichment
+            abundances[u238] = 1.0 - 1.008 * enrichment
+
+            # Convert the mass fractions to mole fractions
+            for i in range(n_nuclides):
+                abundances[i] /= atomic_masses[i]
+
+            # Normalize the mole fractions to one
+            sum_abundances = sum(abundances)
+            for i in range(n_nuclides):
+                abundances[i] /= sum_abundances
+
+        # Compute the ratio of the nuclide atomic massess to the element
+        # atomic mass
+        if percent_type == 'wo':
+
+            # Compute the element atomic mass
+            element_am = 0.
+            for i in range(n_nuclides):
+                element_am += atomic_masses[i] * abundances[i]
+
+            # Convert the molar fractions to mass fractions
+            for i in range(n_nuclides):
+                abundances[i] *= atomic_masses[i] / element_am
+
+            # Normalize the mass fractions to one
+            sum_abundances = sum(abundances)
+            for i in range(n_nuclides):
+                abundances[i] /= sum_abundances
+
+        # Create a list of the isotopes in this element
         isotopes = []
-        for isotope, abundance in sorted(NATURAL_ABUNDANCE.items()):
-            if re.match(r'{}\d+'.format(self.name), isotope):
-                nuc = openmc.Nuclide(isotope)
-                isotopes.append((nuc, abundance))
+        for nuclide, abundance in zip(nuclides, abundances):
+            pct = float('{:2.8g}'.format(percent*abundance))
+            isotopes.append((nuclide, pct, percent_type))
+
```

Comment:
I feel like a lot of this code could be simplified. A few suggestions:

- Don't create `Nuclide` objects until the very last for loop; you can simply keep track of the names.
- Replace `nuclides` and `abundances` lists with a dictionary that maps nuclide names to abundances, e.g. `{'U234': ..., 'U235': ...}`
- Use `mutual_nuclides`, `absent_nuclides`, and `library_nuclides` for looping purposes, e.g. adopting this with the previous suggestions would turn

    ```Python
    if len(absent_nuclides) == 0:
        for nuclide, abundance in sorted(NATURAL_ABUNDANCE.items()):
            if re.match(r'{}\d+'.format(self.name), nuclide):
                nuc = openmc.Nuclide(nuclide)
                nuc.scattering = self.scattering
                nuclides.append(nuc)
                abundances.append(abundance)
    ```
    into
    ```Python
    if len(absent_nuclide) == 0:
        for nuclide in mutual_nuclides:
            abundances[nuclide] = NATURAL_ABUNDANCE[nuclide]
    ```